### PR TITLE
Added Condition shortcut suggestion

### PR DESCRIPTION
**Pull request by bjdmeest**

### CODE REVIEW

**Comment by bjdmeest** [comment]
Current PR needs some more details, but I prefer to have you have a look at the overall idea first before I spend more time

File: spec/resources/example-condition.ttl [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# (isNotNull and IF are defined below, rest is an excercise for the reader)
+<#NameMapping>
+    rml:predicate dbo:title ;
+    # A condition can be defined in any expression map
+    rml:objectMap [
+        # new predicate that links to a function-valued expression map,
+        # that function MUST return a boolean
+        rml:condition [
+            # isNotNull(parameter: X) / definition: X != NULL ? TRUE : FALSE ;
+            rml:function fns:isNotNull ;
```
Comment (dachafra) [inline spec/resources/example-condition.ttl:22]: I think `fns` prefix is not defined

File: spec/resources/example-condition.ttl [hunk 2]
```diff
@@ -0,0 +1,66 @@
+@prefix dbo: <http://dbpedia.org/ontology/> .
+@prefix grel: <http://users.ugent.be/~bjdmeest/function/grel.ttl#> .
```
Comment (dachafra) [inline spec/resources/example-condition.ttl:2]: `grel` is still this URI or should we agree on the new one?

File: spec/docs/advanced.md [hunk 3]
```diff
@@ -64,3 +64,76 @@ For now, it is unclear how to handle a nested function where that nested Triples
...
+                rml:inputValueMap [
+                    rml:reference "name"
+                ]
+            ]
+        ] ;
+        # The actual expression used if the condition returns TRUE
+        rml:constant "true"
+    ] .
+
+# This is actually a shortcut to the following
```
Comment (dachafra) [inline spec/docs/advanced.md:107]: Maybe this part could be in a new snippet

File: spec/docs/advanced.md [hunk 4]
```diff
@@ -64,3 +64,76 @@ For now, it is unclear how to handle a nested function where that nested Triples
...
+@prefix dbo: <http://dbpedia.org/ontology/> .
+@prefix grel: <http://users.ugent.be/~bjdmeest/function/grel.ttl#> .
+@prefix rml: <http://w3id.org/rml/> .
+
+<#Person_Mapping>
+    rml:logicalSource <#LogicalSource> ;
+    rml:subjectMap <#SubjectMap> ;
+    rml:predicateObjectMap <#NameMapping> .
+
+# Suggestion: add rml:condition predicate to expression map,
```
Comment (dachafra) [inline spec/docs/advanced.md:82]: avoid comments inside and move them outside the example?

**Review by dachafra — CHANGES_REQUESTED** [review]

**Review by pmaria — CHANGES_REQUESTED** [review]
So a `rml:condition` is a new property of an ExpressionMap, that if it evaluates to `true` causes the expression map to produce a result, but if `false` causes the expression map to return an empty list.

That in principle sounds good to me.

However, if we use a function execution as value of a condition, one thing that might now cause some problems in the current design of fnml is the separation of `rml:functionExecution` and `rml:return` on `rml:ExpressionMap`. Since `rml:condition` is also placed on the same level, this means that we cannot use `rml:return` for conditions. That seems like a design issue to me.

Perhaps we can solve this by making the value of a condition an expression map as well?


Some specific comments inline

File: spec/resources/example-condition.ttl (hunk 1 above)
Comment (bjdmeest) [reply spec/resources/example-condition.ttl]: fixed

File: spec/resources/example-condition.ttl (hunk 2 above)
Comment (bjdmeest) [reply spec/resources/example-condition.ttl]: using the example.com/fns for now

File: spec/docs/advanced.md (hunk 4 above)
Comment (bjdmeest) [reply spec/docs/advanced.md:82]: fixed

File: spec/resources/turtle.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
+/*
```
[the remark this answers is not in the document]
Comment (bjdmeest) [reply spec/resources/turtle.js]: fixed

File: spec/docs/advanced.md [hunk 6]
```diff
@@ -64,3 +64,76 @@ For now, it is unclear how to handle a nested function where that nested Triples
...
+            rml:input [
+                # The parameter that is checked for NULL
+                rml:parameter fns:parameter ;
+                rml:inputValueMap [
+                    rml:reference "name"
+                ]
+            ]
+        ] ;
+        # The actual expression used if the condition returns TRUE
+        rml:constant "true"
```
[the remark this answers is not in the document]
Comment (bjdmeest) [reply spec/docs/advanced.md:104]: I chose a more fitting constant :)

**Comment by bjdmeest** [comment]
> However, if we use a function execution as value of a condition, one thing that might now cause some problems in the current design of fnml is the separation of `rml:functionExecution` and `rml:return` on `rml:ExpressionMap`. Since `rml:condition` is also placed on the same level, this means that we cannot use `rml:return` for conditions. That seems like a design issue to me.
> 
> Perhaps we can solve this by making the value of a condition an expression map as well?

Yes. I agree and changed accordingly.

**Pull request merged**